src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetectorFactory.cs

```diff
@@ -0,0 +1,99 @@
+namespace Microsoft.ComponentDetection.Detectors.Pnpm;
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.ComponentDetection.Contracts;
+using Microsoft.ComponentDetection.Contracts.Internal;
+using Microsoft.ComponentDetection.Contracts.TypedComponent;
+using Microsoft.Extensions.Logging;
+
+/// <summary>
+/// Factory responsible for constructing the proper <see cref="IPnpmDetector"/> and recording its dependency
+/// graph based on the file found during file component detection.
+/// </summary>
+public class PnpmComponentDetectorFactory : FileComponentDetector
+{
+    public PnpmComponentDetectorFactory(
+        IComponentStreamEnumerableFactory componentStreamEnumerableFactory,
+        IObservableDirectoryWalkerFactory walkerFactory,
+        ILogger<PnpmComponentDetectorFactory> logger)
+    {
+        this.ComponentStreamEnumerableFactory = componentStreamEnumerableFactory;
+        this.Scanner = walkerFactory;
+        this.Logger = logger;
+    }
+
+    public override string Id { get; } = "Pnpm";
+
+    public override IEnumerable<string> Categories => new[] { Enum.GetName(typeof(DetectorClass), DetectorClass.Npm) };
+
+    public override IList<string> SearchPatterns { get; } = new List<string> { "shrinkwrap.yaml", "pnpm-lock.yaml" };
+
+    public override IEnumerable<ComponentType> SupportedComponentTypes { get; } = new[] { ComponentType.Npm };
+
+    public override int Version { get; } = 6;
+
+    public override bool NeedsAutomaticRootDependencyCalculation => true;
+
+    protected override async Task OnFileFoundAsync(ProcessRequest processRequest, IDictionary<string, string> detectorArgs)
+    {
+        var singleFileComponentRecorder = processRequest.SingleFileComponentRecorder;
+        var file = processRequest.ComponentStream;
+        this.Logger.LogDebug("Found yaml file: {YamlFile}", file.Location);
+        var skippedFolder = this.SkippedFolders.FirstOrDefault(folder => file.Location.Contains(folder));
+        if (!string.IsNullOrEmpty(skippedFolder))
+        {
+            this.Logger.LogDebug("Skipping found file, it was detected as being within a {SkippedFolder} folder.", skippedFolder);
+        }
+
+        try
+        {
+            var fileContent = await new StreamReader(file.Stream).ReadToEndAsync();
+            var detector = this.GetPnpmComponentDetector(fileContent, file.Location);
+            if (detector == null)
+            {
+                this.Logger.LogWarning("Unsupported lockfileVersion in pnpm yaml file {File}", file.Location);
+            }
+            else
+            {
+                detector.RecordDependencyGraphFromFile(fileContent, singleFileComponentRecorder);
+            }
+        }
+        catch (Exception e)
+        {
```

Comment:
can you add a `FailedParsingFileRecord` telemetry object here? PipReportComponentDetector has an example